generic_config_updater/patch_sorter.py

```diff
@@ -0,0 +1,965 @@
+            return
+
+        if not isinstance(ptr, dict):
+            for move in self._traverse_target_value(ptr, current_tokens, target_tokens):
+                yield move
+            return
+
+        if len(ptr) == 0:
+            yield JsonMove(self.diff, OperationType.ADD, current_tokens, target_tokens)
+            return
+
+        for key in ptr:
+            current_tokens.append(key)
+            target_tokens.append(key)
+            for move in self._traverse_target(ptr[key], current_tokens, target_tokens):
+                yield move
+            target_tokens.pop()
+            current_tokens.pop()
+
+    def _traverse_target_list(self, ptr, current_tokens, target_tokens):
+        if len(ptr) == 0:
+            yield JsonMove(self.diff, OperationType.ADD, current_tokens, target_tokens)
+            return
+
+        for index, val in enumerate(ptr):
+            # _traverse_target_list is called when the whole list is missing
+            # in such case any item should be added at first location i.e. 0
+            current_tokens.append(0)
+            target_tokens.append(index)
+            for move in self._traverse_target_value(val, current_tokens, target_tokens):
+                yield move
+            target_tokens.pop()
+            current_tokens.pop()
+
+    def _traverse_target_value(self, val, current_tokens, target_tokens):
+        yield JsonMove(self.diff, OperationType.ADD, current_tokens, target_tokens)
+
+    def _list_to_dict_with_count(self, items):
+        counts = dict()
+
+        if items is None:
+            return counts
+
+        for i in items:
+            counts[i] = counts.get(i, 0) + 1
+
+        return counts
+
+    def _check_same_dict_with_count(self, dict1, dict2):
+        if len(dict1) != len(dict2):
+            return False
+        for key in dict1:
+            if key not in dict2:
+                return False
+            if dict1[key] != dict2[key]:
+                return False
+        # No need to check keys in dict2 that are not in dict1.
+        # Because dict1 and dict2 have the same number of keys and dict1 is already checked.
+        # In other words if dict2 has different keys, then for sure dict1 has
+        # also different keys and this will fail in tha above loop.
+        return True
+
+class UpperLevelMoveExtender:
+    def extend(self, move, diff):
+        if not move.current_config_tokens:
+            return
+
+        upper_current_tokens = move.current_config_tokens[:-1]
+        operation_type = self._get_upper_operation(upper_current_tokens, diff)
+
+        upper_target_tokens = None
+        if operation_type in [OperationType.ADD, OperationType.REPLACE]:
+            upper_target_tokens = upper_current_tokens
+
+        yield JsonMove(diff, operation_type, upper_current_tokens, upper_target_tokens)
+
+    # get upper operation assumes ConfigDb to not have list-of-objects, only list-of-values
+    def _get_upper_operation(self, tokens, diff):
```

Comment:
add to the comment, this is especially tricky for remove op as we don't know the corresponding target tokens

Reply:
nvm